### PR TITLE
Cancelling event participants

### LINKED ISSUE #1219: 🤸‍♀️ Cancelling event participants
## Description
A user should be able to mark an event participant as "cancelled", meaning that they have been booked on the event previously, but will not participate.

Also see epic #1214 for context

## Relevant Job Stories
* When I’m organizing an event and a person who previuosly signed up tells me that they will not be able to attend, I want to mark them as cancelled, so that we don’t try to mobilize them for that event

## Prerequisites
Don't start this until #1215 and #1217 have been merged.

This also requires new features in the API that are not implemented yet at the time of writing.

## Requirements
- [ ] The summary card (introduced by #1217) should have a "Cancelled" column, indicating number of cancelled participants
- [ ] At the bottom of the screen, there should be a third section for cancelled participants
  - [ ] The list should mimic the others (sign-ups and booked participants)
  - [ ] The action button on each row in the cancelled participants list should be "Book" (which undos the cancellation)
- [ ] Each item in the "Booked participants" list should have a cancel action button

## Possible implementations
None

### Cancelled count in summary
image

### Cancel buttons in participants list
image

### Cancel section at bottom of screen
image

## Open questions
What will the API look like? See thread below for updates when the API is available.

### PR DESCRIPTION
**Pull request by niklasva82**
## Description
This PR implements cancellation of participants as well as the rebooking of participants.

## Screenshots
[Screencast from 2023-05-04 14-48-36.webm](https://user-images.githubusercontent.com/14962107/236212182-20f10f3b-89ae-494b-a726-b75dd980c39f.webm)

## Changes
* Adds "Cancelled" section to event Participants tab.
* Adds onClick to Cancel button to PUT { status: 'cancelled' } to the EventParticipant
* Adds onClick to Book button in Cancelled Participants section to PUT { status: null }
* Adds number of cancelled participants to the Participant Summary Card.
* When cancelling the event contact, update event in state to set `contact` to `null`

## Notes to reviewer
Backend needs to be updated with zetkin/zetkin-platform#614 before this works properly.

## Related issues
Resolves #1219

### CODE REVIEW

File: src/features/events/components/EventParticipantsList.tsx [hunk 1]
```diff
@@ -42,9 +43,26 @@ const EventParticipantsList: FC<EventParticipantsListProps> = ({
...
         model={model}
         orgId={orgId}
-        rows={model.getParticipants().data ?? []}
+        rows={model.getBookedParticipants()}
         title={messages.eventParticipantsList.bookedParticipants()}
+        type="booked"
       />
+      {model.getNumCancelledParticipants() > 0 && (
+        <ParticipantListSection
+          chipColor={model.getParticipantStatus()}
```
Comment (richardolsson) [inline src/features/events/components/EventParticipantsList.tsx:52]: This should not be dynamically colored, but always be gray (like the signups section).

File: src/features/events/components/EventParticipantsList.tsx [hunk 2]
```diff
@@ -42,9 +43,26 @@ const EventParticipantsList: FC<EventParticipantsListProps> = ({
...
+          chipNumber={
+            model.getNumAvailParticipants() +
+            '/' +
+            data.num_participants_required
+          }
```
Comment (richardolsson) [inline src/features/events/components/EventParticipantsList.tsx:57]: This should be the number of cancelled participants, not identical to the participants section.

File: src/features/events/repo/EventsRepo.ts [hunk 3]
```diff
@@ -226,4 +227,29 @@ export default class EventsRepo {
...
+      .then((participant) => {
+        this._store.dispatch(participantUpdated([eventId, participant]));
+
+        // Get event and check if participant is the contact
+        if (participant.cancelled) {
+          const event = this.getEvent(orgId, eventId).data;
+          if (event?.contact?.id === personId) {
+            this._store.dispatch(eventUpdated({ ...event, contact: null }));
+          }
+        }
```
Comment (richardolsson) [inline src/features/events/repo/EventsRepo.ts:251]: Nice! 💯 

But maybe this could actually be done in the store instead?

**Review by richardolsson — CHANGES_REQUESTED** [review]
Nice work on this, especially thinking about that case where the contact is cancelled. I didn't think about that! 💯 

There are some things that I think could be improved though (comments below), and I would also like you to merge `main` and make sure that the filtering introduced by #1285 also works with this. It should hopefully be quite simple.

File: src/features/events/components/EventParticipantsList.tsx (hunk 2 above)
Comment (niklasva82) [reply src/features/events/components/EventParticipantsList.tsx]: Makes sense, missed that. Fixed.

File: src/features/events/components/EventParticipantsList.tsx (hunk 1 above)
Comment (niklasva82) [reply src/features/events/components/EventParticipantsList.tsx]: Fixed.

**Comment by niklasva82** [comment]
Issues resolved and I merged main and verified that everything worked as expected.

File: src/features/events/store.ts [hunk 4]
```diff
@@ -142,6 +142,34 @@ const eventsSlice = createSlice({
...
+        item.data = { ...item.data, ...participant };
+        item.mutating = [];
+      } else {
+        state.participantsByEventId[eventId].items.push(
+          remoteItem(participant.id, { data: participant })
+        );
+      }
+
+      if (participant.cancelled) {
+        // If cancelled participant was contact for event, also remove contact
```
Comment (richardolsson) [inline src/features/events/store.ts:164]: This condition is not actually checked anywhere in the code. The contact person is just removed regardless of whether it was the participant deleted or not. So there's a bug here that needs to be fixed.

**Review by richardolsson — CHANGES_REQUESTED** [review]
Great work! We're really close to being able to merge this now, but I found a bug in the store code that needs to be fixed. Details below. 👇

File: src/features/events/store.ts (hunk 4 above)
Comment (niklasva82) [reply src/features/events/store.ts:164]: Sorry it went a little too fast there. Fixed!

File: src/features/events/store.ts [hunk 5]
```diff
@@ -142,6 +142,34 @@ const eventsSlice = createSlice({
...
+        const event = state.eventList.items.find(
+          (e) => e?.data?.id === eventId
+        );
+        if (event?.data && event?.data?.contact?.id == participant.id) {
```
Comment (richardolsson) [inline src/features/events/store.ts:168]: It should be enough to just do this:

```suggestion
        if (event?.data?.contact?.id == participant.id) {
```

But no need to change it now. I'm proposing a new lint rule that would have caught this, and when that's introduced there will be a lot of little fixes like this that can all be handled in one go.

**Review by richardolsson — APPROVED** [review]
Nice work!

**Pull request merged**